### PR TITLE
P2300 enhancements

### LINKED ISSUE #5838: [Senders/Receivers] `as_awaitable`
## Expected Behavior

*All awaitables are senders*
Quoting from [P2300](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2300r4.html#spec-execution.coro_utils.as_awaitable) 
```
as_awaitable is used to transform an object into one that is awaitable within a particular coroutine
```
This function needs to be implemented for Senders/Receivers 

## Usage

To make non-awaitable senders as awaitable. 

## Possible cues to implementation

type sender-awaitable as described in the above P2300 and type checking for Senders using templates to check if they have `co_await` method/free operator

CC @hkaiser

### PR DESCRIPTION
**Pull request by SAtacker**
Fixes #5838 

## Proposed Changes

  - Adapt `get_completion_signatures` when Sender is a `awaitable`
  - `as_awaitable_t` with `receiver_base`, `sender_awaitable_base`
  - `promise_base`, `operation_base`
  - `connect_awaitable` to adapt connect mentioned in [spec](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2300r5.html#spec-execution.senders.connect) 2.2
  - `with_awaitable_senders`

## Any background context you want to provide?

- Utility `as_awaitable_t` to transform an object into one that is awaitable within a particular coroutine.
- Utility `with_awaitable_senders` used as the base class of a coroutine promise type, makes senders awaitable in that coroutine type

## Checklist

- [x] I have added a new feature and have added tests to go along with it.

### CODE REVIEW

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 1]
```diff
@@ -703,6 +715,643 @@ namespace hpx::execution::experimental {
...
+            hpx::functional::is_nothrow_tag_invocable_v<as_awaitable_t,
+                Awaitable, promise&> ||
+            !hpx::functional::is_tag_invocable_v<as_awaitable_t, Awaitable,
+                promise&>)
+
+        {
+            if constexpr (hpx::functional::is_tag_invocable_v<as_awaitable_t,
+                              Awaitable, promise&>)
+            {
+                return tag_invoke(
```
[the remark this answers is not in the document]
Comment (hkaiser) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:1193]: Or should it even be: 
```suggestion
                return as_awaitable(HPX_FORWARD(Awaitable, await), *this);
```

File: libs/core/execution_base/include/hpx/execution_base/coroutine_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#include <hpx/execution_base/get_env.hpp>
+#include <hpx/execution_base/traits/coroutine_traits.hpp>
+#include <hpx/functional/detail/tag_fallback_invoke.hpp>
+#include <hpx/type_support/meta.hpp>
+
+#include <exception>
+#include <system_error>
+#include <utility>
+
+#if defined(HPX_HAVE_CXX20_COROUTINES)
```
Comment (hkaiser) [inline libs/core/execution_base/include/hpx/execution_base/coroutine_utils.hpp:21]:
```suggestion
#include <hpx/config.hpp>

#if defined(HPX_HAVE_CXX20_COROUTINES)
#include <hpx/assert.hpp>
#include <hpx/datastructures/variant.hpp>
#include <hpx/execution_base/completion_signatures.hpp>
#include <hpx/execution_base/get_env.hpp>
#include <hpx/execution_base/traits/coroutine_traits.hpp>
#include <hpx/functional/detail/tag_fallback_invoke.hpp>
#include <hpx/type_support/meta.hpp>

#include <exception>
#include <system_error>
#include <utility>

```

**Review by hkaiser — CHANGES_REQUESTED** [review]

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 3]
```diff
@@ -703,6 +715,643 @@ namespace hpx::execution::experimental {
...
+        {
+            using type = A;
+        };
+
+        template <template <typename...> typename Fn>
+        struct compose_template_func
+        {
+            template <typename... Args>
+            using apply = Fn<Args...>;
+        };
```
[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:736]: Alright

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 4]
```diff
@@ -703,6 +715,643 @@ namespace hpx::execution::experimental {
...
+        using single_t =
+            std::enable_if_t<sizeof...(As) == 1, meta::type<front<As...>>>;
+
+        template <typename Ty>
+        struct single_or
+        {
+            template <typename... As>
+            using apply = std::enable_if_t<sizeof...(As) <= 1,
+                meta::type<front<As..., Ty>>>;
+        };
```
[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:748]: yes, thanks

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 5]
```diff
@@ -703,6 +715,643 @@ namespace hpx::execution::experimental {
...
+            Value await_resume()
+            {
+                switch (result.index())
+                {
+                case 0:    // receiver contract not satisfied
+                    HPX_ASSERT_MSG(0, "_Should never get here");
+                    break;
+                case 1:    // set_value
+                    if constexpr (!std::is_void_v<Value>)
+                        return HPX_FORWARD(Value, std::get<1>(result));
```
[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:892]: Can the `result` hold an `rvalue` type?

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 6]
```diff
@@ -703,6 +715,643 @@ namespace hpx::execution::experimental {
...
+                        return hpx::coro::coroutine_handle<
+                            OtherPromise>::from_address(address)
+                            .promise()
+                            .unhandled_stopped();
+                    };
+                }
+                // If OtherPromise doesn't implement unhandled_stopped(),
+                // then if a "stopped" unwind reaches this point,
+                // it's considered an unhandled exception and terminate()
+                // is called.
```
[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:1002]: That's what the standard implementation is doing. Their reasoning convinced me - [ref](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2300r5.html#design-native-coro-unwind)

[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:1002]: But for HPX, I think we can call something other than std::terminate as we even launch our own main function.

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 7]
```diff
@@ -703,6 +715,643 @@ namespace hpx::execution::experimental {
...
+        {
+            return operation<ReceiverId>{
+                hpx::coro::coroutine_handle<promise>::from_promise(*this)};
+        }
+
+        // template <typename Awaitable>
+        // Awaitable&& await_transform(Awaitable&& await) noexcept
+        // {
+        //     return HPX_FORWARD(Awaitable, await);
+        // }
```
[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:1152]: Yes, sorry.

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:1165]: Understood, the latter one seems best suited.

File: libs/core/execution_base/include/hpx/execution_base/coroutine_utils.hpp (hunk 2 above)
Comment (SAtacker) [reply libs/core/execution_base/include/hpx/execution_base/coroutine_utils.hpp:21]: Yes, thanks

File: libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp [hunk 8]
```diff
@@ -703,6 +715,615 @@ namespace hpx::execution::experimental {
...
+            }
+            else
+            {
+                return HPX_FORWARD(T, t);
+            }
+        }
+    } as_awaitable{};
+
+    namespace detail {
+        struct with_awaitable_senders_base
```
Comment (hkaiser) [inline libs/core/execution_base/include/hpx/execution_base/completion_signatures.hpp:977]: Codacy is complaining about this type having private data members but no constructors. I know that this is a false positive and I have tried to convince it to ignore the issue. Could you please add a 
```
    with_awaitable_senders_base() = default;
```
to this class to silence the warning?

**Review by hkaiser — DISMISSED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Pull request merged**